Description: aic_load_fw/bsp: bound ext patch count to the patch table
 Remember the pair count before it is replaced, and clamp ext_patch_nb
 to the pairs that remain after the fixed header. A count that does not
 fit, or a length that cannot describe an allocated table, is not
 walked. Load loops index that list with an unsigned offset.
Author: N Oliver <3945917+PixnBits@users.noreply.github.com>
Forwarded: no
Last-Update: 2026-10-26
---
 .../aic8800/aic8800_bsp/aic_bsp_driver.c      | 86 ++++++++++++++-----
 .../aic8800/aic_load_fw/aic_compat_8800d80.c  |  8 +-
 .../aic_load_fw/aic_compat_8800d80x2.c        |  8 +-
 .../aic8800/aic_load_fw/aicbluetooth.c        | 78 ++++++++++++-----
 4 files changed, 131 insertions(+), 49 deletions(-)

diff --git a/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic_bsp_driver.c b/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic_bsp_driver.c
index e580133..5d5c223 100644
--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic_bsp_driver.c
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_bsp/aic_bsp_driver.c
@@ -1451,39 +1451,77 @@ int aicbt_patch_info_unpack(struct aicbt_patch_info_t *patch_info, struct aicbt_
     uint8_t *patch_info_array = (uint8_t*)patch_info;
     int base_len = 0;
     int memcpy_len = 0;
-    
+    uint32_t table_len;
+    size_t copy_bytes, dst_room;
+
+    if (!patch_info || !head_t)
+        return -1;
+
     if (AICBT_PT_INF == head_t->type) {
         base_len = ((offsetof(struct aicbt_patch_info_t,  ext_patch_nb_addr) - offsetof(struct aicbt_patch_info_t,  adid_addrinf) )/sizeof(uint32_t))/2;
-        AICWFDBG(LOGDEBUG, "%s head_t->len:%d base_len:%d \r\n", __func__, head_t->len, base_len);
+        table_len = head_t->len;
+        AICWFDBG(LOGDEBUG, "%s head_t->len:%d base_len:%d \r\n", __func__, table_len, base_len);
 
-        if (head_t->len > base_len){
+        if (table_len > (uint32_t)base_len){
             patch_info->info_len = base_len;
-            memcpy_len = patch_info->info_len + 1;//include ext patch nb     
+            memcpy_len = patch_info->info_len + 1;//include ext patch nb
         } else{
-            patch_info->info_len = head_t->len;
+            patch_info->info_len = table_len;
             memcpy_len = patch_info->info_len;
         }
 	head_t->len = patch_info->info_len;
-        AICWFDBG(LOGDEBUG, "%s memcpy_len:%d \r\n", __func__, memcpy_len);   
+        AICWFDBG(LOGDEBUG, "%s memcpy_len:%d \r\n", __func__, memcpy_len);
+
+        if (patch_info->info_len == 0) {
+            patch_info->ext_patch_nb = 0;
+            patch_info->ext_patch_param = NULL;
+            return 0;
+        }
 
-        if (patch_info->info_len == 0)
+        /* stop before the ext_patch_param pointer; len is a pair count */
+        dst_room = offsetof(struct aicbt_patch_info_t, ext_patch_param) -
+            sizeof(patch_info->info_len);
+        copy_bytes = (size_t)memcpy_len * sizeof(uint32_t) * 2;
+        if (!head_t->data || (uint32_t)memcpy_len > table_len ||
+            copy_bytes > dst_room || table_len > (~0U / 8)) {
+            patch_info->info_len = 0;
+            patch_info->ext_patch_nb = 0;
+            patch_info->ext_patch_param = NULL;
+            head_t->len = 0;
             return 0;
-       
-        memcpy(((patch_info_array) + sizeof(patch_info->info_len)), 
-            head_t->data, 
-            memcpy_len * sizeof(uint32_t) * 2);
-        AICWFDBG(LOGDEBUG, "%s adid_addrinf:%x addr_adid:%x \r\n", __func__, 
+        }
+
+        memcpy(((patch_info_array) + sizeof(patch_info->info_len)),
+            head_t->data,
+            copy_bytes);
+        AICWFDBG(LOGDEBUG, "%s adid_addrinf:%x addr_adid:%x \r\n", __func__,
             ((struct aicbt_patch_info_t *)patch_info_array)->adid_addrinf,
             ((struct aicbt_patch_info_t *)patch_info_array)->addr_adid);
 
-        if (patch_info->ext_patch_nb > 0){
-            int index = 0;
-            patch_info->ext_patch_param = (uint32_t *)(head_t->data + ((memcpy_len) * 2));
-            
-            for(index = 0; index < patch_info->ext_patch_nb; index++){
-                AICWFDBG(LOGDEBUG, "%s id:%x addr:%x \r\n", __func__, 
-                    *(patch_info->ext_patch_param + (index * 2)),
-                    *(patch_info->ext_patch_param + (index * 2) + 1));
+        if (memcpy_len <= base_len) {
+            /* ext_patch_nb was not in the copied pairs */
+            patch_info->ext_patch_nb = 0;
+            patch_info->ext_patch_param = NULL;
+        } else if (patch_info->ext_patch_nb > 0) {
+            uint32_t nb = patch_info->ext_patch_nb;
+            uint32_t room = table_len - (uint32_t)memcpy_len;
+            uint32_t index = 0;
+
+            if (nb > room)
+                nb = room;
+            if (nb > (~0U / 2))
+                nb = ~0U / 2;
+            patch_info->ext_patch_nb = nb;
+            if (!nb) {
+                patch_info->ext_patch_param = NULL;
+            } else {
+                patch_info->ext_patch_param =
+                    head_t->data + ((size_t)memcpy_len * 2);
+                for (index = 0; index < nb; index++) {
+                    AICWFDBG(LOGDEBUG, "%s id:%x addr:%x \r\n", __func__,
+                        *(patch_info->ext_patch_param + ((size_t)index * 2)),
+                        *(patch_info->ext_patch_param + ((size_t)index * 2) + 1));
+                }
             }
         }
 
@@ -1496,12 +1534,14 @@ int aicbt_ext_patch_data_load(struct aic_sdio_dev *sdiodev, struct aicbt_patch_i
     int ret = 0;
     uint32_t ext_patch_nb = patch_info->ext_patch_nb;
     char ext_patch_file_name[50];
-    int index = 0;
+    uint32_t index = 0;
     uint32_t id = 0;
     uint32_t addr = 0;
 
     
     if (ext_patch_nb > 0){
+        if (!patch_info->ext_patch_param)
+            return -1;
         if (sdiodev->chipid == PRODUCT_ID_AIC8800DC) {
 			AICWFDBG(LOGDEBUG, "[0x40506004]: 0x04318000\n");
 			ret = rwnx_send_dbg_mem_write_req(sdiodev, 0x40506004, 0x04318000);
@@ -1518,8 +1558,8 @@ int aicbt_ext_patch_data_load(struct aic_sdio_dev *sdiodev, struct aicbt_patch_i
 			ret = rwnx_send_dbg_mem_write_req(sdiodev, 0x40509000, 0x00040000);
         }
         for (index = 0; index < patch_info->ext_patch_nb; index++){
-            id = *(patch_info->ext_patch_param + (index * 2));
-            addr = *(patch_info->ext_patch_param + (index * 2) + 1); 
+            id = *(patch_info->ext_patch_param + ((size_t)index * 2));
+            addr = *(patch_info->ext_patch_param + ((size_t)index * 2) + 1); 
             memset(ext_patch_file_name, 0, sizeof(ext_patch_file_name));
             sprintf(ext_patch_file_name,"%s%d.bin",
                 aicbsp_firmware_list[aicbsp_info.cpmode].bt_ext_patch,
diff --git a/src/USB/driver_fw/drivers/aic8800/aic_load_fw/aic_compat_8800d80.c b/src/USB/driver_fw/drivers/aic8800/aic_load_fw/aic_compat_8800d80.c
index 303bf57..02e9480 100644
--- a/src/USB/driver_fw/drivers/aic8800/aic_load_fw/aic_compat_8800d80.c
+++ b/src/USB/driver_fw/drivers/aic8800/aic_load_fw/aic_compat_8800d80.c
@@ -373,16 +373,18 @@ static int aicbt_ext_patch_data_load(struct aic_usb_dev *usb_dev, struct aicbt_p
     int ret = 0;
     uint32_t ext_patch_nb = patch_info->ext_patch_nb;
     char ext_patch_file_name[50];
-    int index = 0;
+    uint32_t index = 0;
     uint32_t id = 0;
     uint32_t addr = 0;
 
 
     if (ext_patch_nb > 0){
+        if (!patch_info->ext_patch_param)
+            return -1;
 
         for (index = 0; index < patch_info->ext_patch_nb; index++){
-            id = *(patch_info->ext_patch_param + (index * 2));
-            addr = *(patch_info->ext_patch_param + (index * 2) + 1);
+            id = *(patch_info->ext_patch_param + ((size_t)index * 2));
+            addr = *(patch_info->ext_patch_param + ((size_t)index * 2) + 1);
             memset(ext_patch_file_name, 0, sizeof(ext_patch_file_name));
             sprintf(ext_patch_file_name,"%s%d.bin",
                 filename,
diff --git a/src/USB/driver_fw/drivers/aic8800/aic_load_fw/aic_compat_8800d80x2.c b/src/USB/driver_fw/drivers/aic8800/aic_load_fw/aic_compat_8800d80x2.c
index e74d608..ab2e472 100644
--- a/src/USB/driver_fw/drivers/aic8800/aic_load_fw/aic_compat_8800d80x2.c
+++ b/src/USB/driver_fw/drivers/aic8800/aic_load_fw/aic_compat_8800d80x2.c
@@ -396,13 +396,15 @@ static int aicbt_ext_patch_data_load(struct aic_usb_dev *usb_dev, struct aicbt_p
     int ret = 0;
     uint32_t ext_patch_nb = patch_info->ext_patch_nb;
     char ext_patch_file_name[50];
-    int index = 0;
+    uint32_t index = 0;
     uint32_t id = 0;
     uint32_t addr = 0;
     uint32_t mem_w_add = 0;
     uint32_t mem_w_data = 0;
 
     if (ext_patch_nb > 0){
+        if (!patch_info->ext_patch_param)
+            return -1;
         AICWFDBG(LOGDEBUG, "[0x40480000]: 0x00040220\n");
         mem_w_add = 0x40580000;
         mem_w_data = 0x00040220;
@@ -414,8 +416,8 @@ static int aicbt_ext_patch_data_load(struct aic_usb_dev *usb_dev, struct aicbt_p
         }
 
         for (index = 0; index < patch_info->ext_patch_nb; index++){
-            id = *(patch_info->ext_patch_param + (index * 2));
-            addr = *(patch_info->ext_patch_param + (index * 2) + 1);
+            id = *(patch_info->ext_patch_param + ((size_t)index * 2));
+            addr = *(patch_info->ext_patch_param + ((size_t)index * 2) + 1);
             memset(ext_patch_file_name, 0, sizeof(ext_patch_file_name));
             sprintf(ext_patch_file_name,"%s%d.bin",
                 filename,
diff --git a/src/USB/driver_fw/drivers/aic8800/aic_load_fw/aicbluetooth.c b/src/USB/driver_fw/drivers/aic8800/aic_load_fw/aicbluetooth.c
index 7e52647..69473d7 100644
--- a/src/USB/driver_fw/drivers/aic8800/aic_load_fw/aicbluetooth.c
+++ b/src/USB/driver_fw/drivers/aic8800/aic_load_fw/aicbluetooth.c
@@ -1481,39 +1481,77 @@ int aicbt_patch_info_unpack(struct aicbt_patch_info_t *patch_info, struct aicbt_
     uint8_t *patch_info_array = (uint8_t*)patch_info;
     int base_len = 0;
     int memcpy_len = 0;
-    
+    uint32_t table_len;
+    size_t copy_bytes, dst_room;
+
+    if (!patch_info || !head_t)
+        return -1;
+
     if (AICBT_PT_INF == head_t->type) {
         base_len = ((offsetof(struct aicbt_patch_info_t,  ext_patch_nb_addr) - offsetof(struct aicbt_patch_info_t,  adid_addrinf) )/sizeof(uint32_t))/2;
-        AICWFDBG(LOGDEBUG, "%s head_t->len:%d base_len:%d \r\n", __func__, head_t->len, base_len);
+        table_len = head_t->len;
+        AICWFDBG(LOGDEBUG, "%s head_t->len:%d base_len:%d \r\n", __func__, table_len, base_len);
 
-        if (head_t->len > base_len){
+        if (table_len > (uint32_t)base_len){
             patch_info->info_len = base_len;
-            memcpy_len = patch_info->info_len + 1;//include ext patch nb     
+            memcpy_len = patch_info->info_len + 1;//include ext patch nb
         } else{
-            patch_info->info_len = head_t->len;
+            patch_info->info_len = table_len;
             memcpy_len = patch_info->info_len;
         }
 	head_t->len = patch_info->info_len;
-        AICWFDBG(LOGDEBUG, "%s memcpy_len:%d \r\n", __func__, memcpy_len);   
+        AICWFDBG(LOGDEBUG, "%s memcpy_len:%d \r\n", __func__, memcpy_len);
+
+        if (patch_info->info_len == 0) {
+            patch_info->ext_patch_nb = 0;
+            patch_info->ext_patch_param = NULL;
+            return 0;
+        }
 
-        if (patch_info->info_len == 0)
+        /* stop before the ext_patch_param pointer; len is a pair count */
+        dst_room = offsetof(struct aicbt_patch_info_t, ext_patch_param) -
+            sizeof(patch_info->info_len);
+        copy_bytes = (size_t)memcpy_len * sizeof(uint32_t) * 2;
+        if (!head_t->data || (uint32_t)memcpy_len > table_len ||
+            copy_bytes > dst_room || table_len > (~0U / 8)) {
+            patch_info->info_len = 0;
+            patch_info->ext_patch_nb = 0;
+            patch_info->ext_patch_param = NULL;
+            head_t->len = 0;
             return 0;
-       
-        memcpy(((patch_info_array) + sizeof(patch_info->info_len)), 
-            head_t->data, 
-            memcpy_len * sizeof(uint32_t) * 2);
-        AICWFDBG(LOGDEBUG, "%s adid_addrinf:%x addr_adid:%x \r\n", __func__, 
+        }
+
+        memcpy(((patch_info_array) + sizeof(patch_info->info_len)),
+            head_t->data,
+            copy_bytes);
+        AICWFDBG(LOGDEBUG, "%s adid_addrinf:%x addr_adid:%x \r\n", __func__,
             ((struct aicbt_patch_info_t *)patch_info_array)->adid_addrinf,
             ((struct aicbt_patch_info_t *)patch_info_array)->addr_adid);
 
-        if (patch_info->ext_patch_nb > 0){
-            int index = 0;
-            patch_info->ext_patch_param = (uint32_t *)(head_t->data + ((memcpy_len) * 2));
-            
-            for(index = 0; index < patch_info->ext_patch_nb; index++){
-                AICWFDBG(LOGDEBUG, "%s id:%x addr:%x \r\n", __func__, 
-                    *(patch_info->ext_patch_param + (index * 2)),
-                    *(patch_info->ext_patch_param + (index * 2) + 1));
+        if (memcpy_len <= base_len) {
+            /* ext_patch_nb was not in the copied pairs */
+            patch_info->ext_patch_nb = 0;
+            patch_info->ext_patch_param = NULL;
+        } else if (patch_info->ext_patch_nb > 0) {
+            uint32_t nb = patch_info->ext_patch_nb;
+            uint32_t room = table_len - (uint32_t)memcpy_len;
+            uint32_t index = 0;
+
+            if (nb > room)
+                nb = room;
+            if (nb > (~0U / 2))
+                nb = ~0U / 2;
+            patch_info->ext_patch_nb = nb;
+            if (!nb) {
+                patch_info->ext_patch_param = NULL;
+            } else {
+                patch_info->ext_patch_param =
+                    head_t->data + ((size_t)memcpy_len * 2);
+                for (index = 0; index < nb; index++) {
+                    AICWFDBG(LOGDEBUG, "%s id:%x addr:%x \r\n", __func__,
+                        *(patch_info->ext_patch_param + ((size_t)index * 2)),
+                        *(patch_info->ext_patch_param + ((size_t)index * 2) + 1));
+                }
             }
         }
 
